### PR TITLE
editor: toggle block style refinements

### PR DESCRIPTION
**Pull request by tvanderstad**
fixes https://github.com/lockbook/lockbook/issues/2034
fixes https://github.com/lockbook/lockbook/issues/2035

https://github.com/lockbook/lockbook/assets/6198756/3fdde85d-9cdb-4982-af21-4ecadb7ee66a

### CODE REVIEW

File: libs/editor/egui_editor/src/input/canonical.rs [hunk 1]
```diff
@@ -93,10 +93,6 @@ pub enum Modification {
     ToggleDebug,
     ToggleCheckbox(usize),
     OpenUrl(String),
-    Heading(u32),
-    BulletListItem,
-    NumberListItem,
-    TodoListItem,
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/input/canonical.rs:99]: no more one-off modification variants for block styles; these are handled by the generic apply_style fn now

File: libs/editor/egui_editor/src/input/canonical.rs [hunk 2]
```diff
@@ -236,17 +232,47 @@ pub fn calc(
...
             if modifiers.command && modifiers.shift =>
         {
-            Some(Modification::BulletListItem)
+            Some(Modification::toggle_block_style(BlockNode::ListItem(ListItem::Bulleted, 0)))
         }
         Event::Key { key: Key::Num9, pressed: true, modifiers, .. }
             if modifiers.command && modifiers.shift =>
         {
-            Some(Modification::TodoListItem)
+            Some(Modification::toggle_block_style(BlockNode::ListItem(ListItem::Todo(false), 0)))
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/input/canonical.rs:245]: additional hotkeys

File: libs/editor/egui_editor/src/input/mutation.rs [hunk 3]
```diff
@@ -59,14 +58,13 @@ pub fn calc(
...
-                buffer,
-                ast,
-                &mut mutation,
-            );
+            let unapply = region_completely_styled(cursor, &style, ast);
+            if !unapply {
+                for conflict in conflicting_styles(cursor, &style, ast) {
+                    apply_style(cursor, conflict, true, buffer, ast, &mut mutation)
+                }
+            }
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/input/mutation.rs:66]: there's now a concept of two styles 'conflicting'; when applying a style, any conflicting styles applied to the same region are removed

File: libs/editor/egui_editor/src/style.rs [hunk 4]
```diff
@@ -56,13 +55,15 @@ impl MarkdownNodeType {
...
-            }
+            Self::Block(BlockNodeType::Code) => "\n```",
             Self::Block(BlockNodeType::ListItem(..)) => "",
             Self::Block(BlockNodeType::Rule) => "",
         }
     }
+
+    pub fn conflicts_with(&self, other: &MarkdownNodeType) -> bool {
+        matches!((self, other), (Self::Block(..), Self::Block(..)))
+    }
```
Comment (tvanderstad) [inline libs/editor/egui_editor/src/style.rs:66]: as a good first pass, any block style conflicts with any other block style

**Review by Parth — APPROVED** [review]

**Pull request merged**